.github/workflows/ci.yml

```diff
@@ -0,0 +1,105 @@
+name: CI
+
+on:
+  push:
+    branches:
+      - main
+  pull_request:
+    branches:
+      - main
+
+jobs:
+  test:
+    name: Test on Python ${{ matrix.python-version }}
+    runs-on: ubuntu-latest
+    permissions:
+      contents: read
+    strategy:
+      matrix:
+        python-version: ["3.11", "3.12"]
+    
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v4
+      
+      - name: Set up Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v5
+        with:
+          python-version: ${{ matrix.python-version }}
+          cache: 'pip'
+      
+      - name: Install dependencies
+        run: |
+          python -m pip install --upgrade pip
+          pip install -e ".[dev]"
+      
+      - name: Run linter
+        run: |
+          ruff check src tests
+        continue-on-error: true
+      
+      - name: Run type checker
+        run: |
+          mypy src
+        continue-on-error: true
+      
+      - name: Run unit tests
+        run: |
+          pytest --cov=ez_agent --cov-report=xml --cov-report=term
+      
+      - name: Upload coverage to Codecov
+        uses: codecov/codecov-action@v5
+        if: matrix.python-version == '3.12'
+        with:
+          file: ./coverage.xml
+          fail_ci_if_error: false
+  
+  cli-validation:
```

Comment:
can this depends on previous job to make sure that we don't run it if the tests does not pass. 
Also can we shared the pip install accross the stage to save time and resources ?